### PR TITLE
Sharing groups

### LINKED ISSUE #90: Support bidirectional sync with a different database file
## Expected Behavior

Option to Sync (on changes) from another external file.

Would be nice if with the current database Autoreload and database Merge we should implement an option on KeePassXC to watch on an external source (a file placed on another drive/folder) and when there are changes on that file, merge with the current opened database.

This can be useful with Cloud/Sync client like Dropbox, GDrive, WebDav, so one can listen on an external file and KeePassXC will continuously merge the changes from that file.

This is related to https://github.com/keepassxreboot/keepassxc/issues/22

### LINKED ISSUE #637: Bidirectional Merge/Synchronization
## Expected Behavior
Bidirectional merge/synchronization

Does the current implementation support *bidirectional* merges/synchronization? If for instance, you store a local copy and a Dropbox copy separately (both available locally, though) and open the local copy and add a new entry, is there a way to merge the local copy's new entry into the Dropbox file (as well as pulling in any changes that are present in the Dropbox file but not yet the local copy, in either one or two steps)? I didn't see it anywhere in the menus and merge only seemed to affect the open database rather than both.

Asking due to the situation noted here http://keepass.info/help/kb/trigger_examples.html in the "Synchronizing with Dropbox..." section (sorry I can't link to that section specifically). At some point I had an issue with Dropbox overwriting with an older copy and dropping some of the changes (and adding a conflict file).

I saw a few issues that were related, but none seemed to address this directly. Apologies if this is a duplicate.

Final question, though it might belong in its own issue, is it possible for merges to assume both databases use the same password&/key and to only ask if the password&/key don't match?

### LINKED ISSUE #2062: Allow user to define automatic DB backup location and name
## Expected Behavior

Allow users to define location (for example this allows saving to a different drive and useful if the drive with the original DB file fails) and also a custom name for the backed up DB file.

## Current Behavior

Currently, there is an option to save a backup before saving - which saves to the same folder location as the original DB file with name as `db.old.kdbx`

### PR DESCRIPTION
**Pull request by ckieschnick**
In many cases passwords should be kept secret and should not be shared with anyone, sometimes it is tiresome to give everybody access to the specific passwords without publishing the rest. Since keeping multiple databases does not do any good, especially when many people may need to access and update them, we created an extension which allows to share groups of entries with multiple peers while keeping others private without the need of multiple databases and regardless of the file location.

## Description
The extension allows to export/import or synchronize(=import+export) groups with their contained entries to other parties. The shared data is placed in a share container which may be published by any means (i.e. cloud, network share, ...). Any client can import this container as long he has the password. 
The sharing container is protected using a certificate which at least ensures that the source is the expected one, so the client may decline to import data from an untrusted source.
The features does make use of the synchronization (#1992) and merges data in time order (newest data will be on top, older data will be moved to the history). 
All data required to work for KeeShare is stored either in KeePassXC-Settings or `CustomData`, so the extension is fully compatible with existing databases.

## Motivation and context
It is often necessary to publish certain secrets to a selected group of people but protecting those secret none the less. Just imagine a company which wants to distribute passwords to some employee groups (i.e. banking data just for the managers, root passwords for admins). Those passwords can be published on internal sites, but keeping them in a password manager makes it easy to protect them. If the password manager takes care, that the passwords are up to date, even when it is not centralized, it would be even better. That's why we introduced the group sharing feature (KeeShare) to KeePassXC. There is no need to keep multiple databases around which may have problems when shared using some technologies. 

Furthermore, the extension addresses some more wishes and improvements from the issue list:
* import features resolves #90 Option to Sync (on changes) from another external file (just make the root group an import group)
* export feature resolves #2062 Allow user to define automatic DB backup location and name (just make the root group an export group and specify the location - to open the backup, one has to import it using the password)
* the merge may be used feature to resolve #637 (Bidirectional Merge/Synchronization)

Problems and Tickets which may gain in importance:
* sharing without history support may lead to data loss
* sharing with autosave on change prevents the user from reviewing and discarding the changes of an import and may lead to unnecessary exports
* #1162 (Provide a visual indication for entries that contain unsaved changes) may make the change tracking of imports easier (depends on the visualization of changes during the import)
* #852 (User should be warned before deleting entry with a UID that is being REF'd by another object)/#1744 (Warn user if deleting entries that are referenced.) may improve sharing since references would be used to export an entry to multiple shares

Possible use cases are:
* bidirectional sharing between different individuals or groups without the need of maintaining several databases at the same time (i.e. using a network share or dropbox)
* simple backup of parts of a database or the whole database to a specific location (using only the export feature)
* unidirectional sharing of specific entries (using export on the master and import on the slave client) There may be more use cases for the sync feature which may help with certain restrictions/adaptions (i.e. auto import of a CSV file into a group - may need an adaption of the merge logic to reduce duplicates)

## How has this been tested?
There are extensions to the test suite, but a lot of functionality is tested manually. I'm note sure if a UI-test would be justified. 

## Screenshots (if appropriate):
Please look in the updated [QUICKSTART](https://github.com/hicknhack-software/keepassxc/blob/feature/sharing_groups/docs/QUICKSTART.md)

## Types of changes
- ✅ New feature (non-breaking change which adds functionality)

## Checklist:
- ✅ I have read the **CONTRIBUTING** document. 
- ✅ My code follows the code style of this project. 
- ✅ All new and existing tests passed.
- ✅ I have compiled and verified my code with `-DWITH_ASAN=ON`.
- ✅ My change requires a change to the documentation and I have updated it accordingly.
- ✅ I have added tests to cover my changes. - As far as the current abstraction allowed.

### CODE REVIEW

**Comment by phoerious** [comment]
Hey, thanks for your follow-up PR. Could you please check why the tests are failing and resolve the merge conflicts? Then we can start reviewing your changes.

**Comment by ckieschnick** [comment]
CodeFactor fails currently due to an intended duplication in `FileDialog`.  I'm not sure how it was intended to use the methods, therefore I didn't try to put my changes into an existing method to prevent unintended changes at other places. 
TeamCity has just one issue (for now) - `Release [GCC, AppImage]` couldn't fetch packages from a specific ubuntu repository - seems to be a transient failure to me. The other images seem to work now.

**Comment by droidmonkey** [comment]
That means we need to increment the version number in the docker script so it rebuilds the image.

**Comment by droidmonkey** [comment]
@ckieschnick sorry for the incredible delay on this PR. We had to clean house a little first. I will rebase this after we merge your first PR. Any additional comments?

**Comment by ckieschnick** [comment]
I'm currently quite busy, so I couldn't track the progress - therefore I have nothing to comment. In case problems arise, I'll try to support you.

**Comment by ratcashdev** [comment]
Please consider adding the option of keeping a configurable number of backups (e.g. 3 older backups), not just one.

**Comment by droidmonkey** [comment]
Some issues I found right away:

 * When adding sync to a group, all the other groups were instantly deleted
 * It is not possible to sync two folders in the same database with each other
 * Inconsistent syncing occurs when two folders point to the same kdbx.share, they should be independent. One entry went to one folder, the others went to another folder.
 * Qua Zip needs to be removed. The container certificate can easily be stored within the kdbx metadata itself. No need for a separate rider file or zipping anything.

**Comment by ckieschnick** [comment]
For clarification: 
* The feature is not intended to sync groups within a database but to sync groups between different databases. It may be possible to have a master slave setup with one group exporting while the slave do only import, but this is more a side effect than an intended feature.
* "Merging" multiple groups into a single kdbx.share (using sync or export) is no use case - each group overwrites the share therefore exporting multiple times from within one database into the same share would lead to lost updates.
* We used QuaZip for simplicity, since the container format is more easy to read and validate (no need to read anything of the container). Another advantage is, that it is easy to use the unpacked container as backup and even checking the signature with third party tools would be easier when the signature is not stored within the container.

Regarding your first point - I didn't have time to check the implementation against the current develop in the last month, but this problem seems weird. I'm currently quite busy, but I may be able to help, if you can provide some more information on this issue.

**Comment by droidmonkey** [comment]
@ckieschnick no problems, these were more notes for me to investigate further and provide fixes. It is very very likely we are going to drop the QuaZip usage, it currently does not have a port for Qt5 on Trusty and possibly not on macOS either.

File: INSTALL.md [hunk 1]
```diff
@@ -98,15 +99,17 @@ These steps place the compiled KeePassXC binary inside the `./build/src/` direct
 	  -DWITH_XC_YUBIKEY=[ON|OFF] Enable/Disable YubiKey HMAC-SHA1 authentication support (default: OFF)
 	  -DWITH_XC_BROWSER=[ON|OFF] Enable/Disable KeePassXC-Browser extension support (default: OFF)
 	  -DWITH_XC_NETWORKING=[ON|OFF] Enable/Disable Networking support (favicon download) (default: OFF)
-	  -DWITH_XC_SSHAGENT=[ON|OFF] Include SSH agent support. (default: OFF)
-	  
+	  -DWITH_XC_SSHAGENT=[ON|OFF] Enable/Disable SSHAgent support (default: OFF)
+	  -DWITH_XC_SHARING=[ON|OFF] Enable/Disable Sharing extension (default: OFF)
```
Comment (TheZ3ro) [inline INSTALL.md]: This is called `WITH_XC_KEESHARE` everywhere else

**Comment by dsonck92** [comment]
I've been testing this feature today. Very good functionality. However, the first thing which struck me as odd (even though it's mentioned in the docs) is the flat initial syncing.

- While I like the idea of being able to move the items outside of the sync group while maintaining the sync, this should be visually represented in a way. I currently have no "dumb trust" it will remain connected with the sync container.
- Another thing. When I first used it, and it synced everything without hierarchy, I first thought it didn't sync the subitems. I later realized it did in fact sync over, but as I have a huge but organized collection of passwords, it basically makes the first sync very messy. All items are merged into the top group. That means I need to recreate the full structure on the other sync. It might be useful to include a checkmark "Sync subgroup structure". To allow both syncing with and without preserving the paths of the keys. For those that do want to keep the hierarchy synced as well.

**Comment by droidmonkey** [comment]
@dsonck92 great points, this does require a little bit of work. I think the default behavior should be to maintain structure, it does not make any sense to eliminate structure that is created.

**Comment by ckieschnick** [comment]
You have a valid point, @dsonck92. We thought about maintaining structure, but this requires a lot more housekeeping and the edge cases (deletion, moving entries) need to be defined very carefully to prevent unexpected behavior (keep in mind that the whole sync should also allow to sync between different users with different structures). 
That's why we implemented just flat sync - a more sophisticated sync strategy like structured sync was thought as nice-to-have.

**Comment by AlexanderAmelkin** [comment]
I'm currently using Keepass2Android on my phone and KeepassXC on two Linux desktops. The database is stored on WebDAV. The Keepass2Android client has native support for WebDAV. For KeepassXC on Linux machines I use WebDAV mounting by the OS. The problem I'm having is that KeepassXC always opens the database in read-only mode and I then have to "Save as..." and overwrite the file when I do any changes on the desktop. This is annoying. Is this addressed in any way by this PR ?

**Comment by dsonck92** [comment]
> I'm currently using Keepass2Android on my phone and KeepassXC on two Linux desktops. The database is stored on WebDAV. The Keepass2Android client has native support for WebDAV. For KeepassXC on Linux machines I use WebDAV mounting by the OS. The problem I'm having is that KeepassXC always opens the database in read-only mode and I then have to "Save as..." and overwrite the file when I do any changes on the desktop. This is annoying. Is this addressed in any way by this PR ?

It provides an alternative way to sync your data but I don't know about support in Keepass2Android for this feature. I currently use a desktop with this feature enabled to create a "phone" database, which gets synced to Nextcloud (which is also supported by Keepass2Android). This gets the synced entries into my phone.

As I understand the end feature, it would solve your problem:
- You would store the "share group" in webdav
- you take the current database out of your webdav folder
- and keepass2android should also use such a sync.

Each would automatically update and import changes.

**Comment by droidmonkey** [comment]
Please don't post issues to this PR. Thank you.

**Comment by dsonck92** [comment]
@skieschnick understandable. I would suggest keeping the flat structure for the time being as the sync feature is very useful even without the structure. But

I highly suggest visualizing the syncing of items:

- whether an item is synced at all
- which direction it is syncing
- what share group it belongs to

That way, it might prevent accidental editing of incoming syncs which get overwritten. Prevent accidental unsynchronized items without noticing it broke the sync.

I haven't delved into the details of how it exactly syncs and when it will still sync items, but an end user should not be expected to do that either. Although I do like the idea of still syncing even outside the share group, as it makes the structure completely independent from others.

As another thought for a new edition which would really be far in the future:

If entries don't need to be under their actual group, why not actually create some sort of tagging system which puts all items tagged as something into such a share group. Of course this would require quite a bit of changes to keepassxc as a whole. As it too might benefit from a tagging system. But that's me going way beyond.

So in short: I highly recommend giving feedback on how the items are synced. This might be an extra column with the name of the group+icon for direction. Just for that visual feedback when items are not underneath their sync group

**Comment by ckieschnick** [comment]
Since this PR is pending for a while now, it would be nice to know which points prevent accepting the code (excluding the merge conflicts). Can you please name the issues so we can address them?

**Comment by droidmonkey** [comment]
Please see my comments from Oct 2 above. I think phase 1 of this PR should not include the quazip library (or be a compile option that is disabled by default). I do not want to add another dependency prior to shipping 2.4.0.

**Comment by droidmonkey** [comment]
I am working on this next week, but as you can see there are a TON of conflicts due to the database/databasewidget refactor that @phoerious performed.

**Comment by ckieschnick** [comment]
Thanks - unfortunately I'm currently busy at work, so there is currently no spare time to fix the conflicts. 
As for the removal of the quazip dependency, we suggest the following: 

- Building without the zipping tooling, we suggest to export a shared database only without any additional files (to allow easy sharing using file sync) and dropping the verification part (in fact using the database as unverified source). 
- Building with quazip would allow to share verified databases. Furthermore it should be able to import share containers of builds without quazip (handling them as unverified sources requesting the allowance for import only the first time).

**Comment by droidmonkey** [comment]
I've picked this up again and did the painful merge up to HEAD.

**Comment by droidmonkey** [comment]
Sorry @ckieschnick I had already done a new develop merge locally and hadn't pushed it yet. Please accept these new commits which incorporates your signal/slot fix. I am working on this today and tomorrow to cleanup the UX and make quazip optional.

**Comment by ckieschnick** [comment]
@droidmonkey today I had the time to update my working copy and made some progress. Most of the work to make quazip an optional dependency is done - it just needs some testing and maybe some polishing before pushing it. Please wait for my changes - I'll push them tomorrow as soon as possible.

**Comment by droidmonkey** [comment]
Ok perfect thank you

**Comment by ckieschnick** [comment]
d4c391d does split the KeeShare functions into a secure and a insecure variant. Only the secure variant needs the quazip path, so simple sharing (using standard database containers) is possible using only the insecure variant. Next, I'll do some testing regarding changed functionality and I like to introduce a better handling for unsigned imports (both for the secure and insecure variant).

**Comment by ckieschnick** [comment]
Quazip is now an optional dependency. KeeShare is activated enabling the KeeShare-feature. In case no Quazip is found, KeeShare exports simple databases, If Quazip was found, KeeShare allows and defaults to secured exports. 

For insecure KeeShare, it is not possible to sync with secured containers but for secure KeeShare it is possible to sync with insecure containers. The default for secure builds is to use secured containers. 

To reduce the risk of unwanted imports, we added a scope to the certificates which is bound to the share container path - this way only accepted containers are automatically synced. 

Concerning the complexity fails of CodeFactor, you may have a look, but I think these cases are tolerable since more separation does not ease reading or understanding of the code.

**Comment by droidmonkey** [comment]
I made a couple minor edits to fix a couple bugs. Here are my observations:

* There is no way to build insecure-only when QuaZip is installed. My opinion is that building KeeShare secure should be opt-in, not based on the presence of libraries. Propose eliminating WITH_XC_KEESHARE_INSECURE variable since WITH_XC_KEESHARE implies this state. Make WITH_XC_KEESHARE_SECURE an option variable that is NOT enabled when WITH_XC_ALL=ON. When building you have to explicitly define WITH_XC_KEESHARE_SECURE=ON which then enables the QuaZip dependency.
* There is no way to test that the sync is working until you press OK to accept the group settings. Even then, when no settings are changed it does not trigger a re-sync which makes it difficult to determine if there is an issue. Suggest a persistent banner that shows when viewing a group that is synced which identifies the group as synced and its current status.
* There is no way to import a certificate to trust. I think a button was accidentally left off the refactor.
* Notices in the group edit should persist and not include a close button.
* Received multiple popup warnings about insecure containers. Even one warning seems excessive considering I know its "insecure".
* I do not agree with the terminology of "insecure" vs "secure" since that implied the database file itself is not secure. Its just "unsigned" or "untrusted". Plain kdbx files should not have any qualifiers and signed shares should be called "Signed KDBX Share *.kdbx.share" or something similar.

**Comment by ckieschnick** [comment]
I'll try to address the points in the same order as given:
* Your proposal using KEESHARE and KEESHARE_SECURE is okay for me. I'll adapt it.
* What do you want to check? For me, the import seems to work as soon as the configuration for the group is accepted. The activated sharing on a group is indicated by the small badge and the "share" tab in the group details.
* I'm not sure which button do you mean. 
* Persistent warnings - seems better. I'll look into it.
* Seems like a bug to me, but I'm have no idea how you were able to create multiple warnings for the same Sync-Container since the scope seems to be identical.
* I'm open for suggestions since I'm not happy with terminology too. Secure means that beside the password the source of the container is trusted. Insecure means, that the password matches, but the source cannot be verified so you have to trust the container blind. As long as only trusted people can access the shared filesystem where container resides, it is not insecure - but in an "open" scenario where anybody may read or write the shared container, one may import data blindly which may lead to undesired results (i.e. a changed entry with enough history to remove any valid data, so the data is lost).
* I forgot to check the MSYS-build.

**Comment by ckieschnick** [comment]
I could reproduce the problem with the multiple dialogs - it's a problem with the `FileWatcher` which does signal a file change multiple times, but it is expected to be send exactly once. The quick and dirty way would be to delay the permission for some milliseconds to lunch the dialog after all signals were emitted, but I'll try to locate the root cause for now.

**Comment by droidmonkey** [comment]
Awesome, I tested everything out and it works much better. I think we can do a code review and merge it in afterwards. Any minor ui/ux issues can be cleaned up in a new PR.

**Comment by droidmonkey** [comment]
@ckieschnick great work on this, it is now a part of 2.4.0

**Comment by Guite** [comment]
Can someone please briefly explain the essential difference of this new addition and the AutoOpen functionality? If I understood the quickstart document correctly there is an additional `kdbx` file here, too.

**Comment by droidmonkey** [comment]
AutoOpen keeps the database OUTSIDE of the currently used database. KeeShare brings the database INSIDE to the currently used database. Also the ability to IMPORT, EXPORT, SYNCHRONIZE versus just SYNCHRONIZE with AutoOpen.

**Comment by Guite** [comment]
Thank you :+1:

**Comment by AlexanderAmelkin** [comment]
So, do I understand it right that you actually can not share the database with other versions of Keepass (e.g. Keepass2Android) and the implementation is essentially useless for mobile users because there is no KeePassXC for Android and no other keepass client has support for this sophisticated mechanism?

**Comment by droidmonkey** [comment]
You cannot update the share (currently) with KeePass2Android. However, the main database you use will have all shared entries stored within it and are accessible by other keepass clients. KeePass2 has support for the KeeShare implementation via a plugin.

**Comment by hi-ko** [comment]
The mentioned quickstart guide is no longer available (404) and should be updated to [QUICKSTART.md](https://github.com/hicknhack-software/keepassxc/blob/develop/docs/QUICKSTART.md)

I also tested the share functionallity in my family and we ended up with a lot of confusion: duplicate entries, keys are not moved to new locations, users do not get when and why entries stop syncrhonization. I think to make "Sharing groups" working the way users expect would require more work than implementing a real sync like available in the C# keepass which is working quite good.
If following the "Sharing groups" aproach the node should contain visible metadata and maybe a visible indicator that it is a shared node and if it is "in sync" - otherwise I wouldn't trust this implementation with my testing experience.

In long term I would prefer a real database approach (postgres, mysql) for managing shared passwords beeing able to define group permissions, audit logs. This could be implemented as a simple and secure REST service in the first step and extended by messaging (e.g. rabbitmq) to "push" changes to the clients while they're connected. I'm willing to provice the server part with my company but we don't have any c skills for the client part.

**Comment by droidmonkey** [comment]
Please open a new issue (search first please) for any suggestions or bugs you come across. Please don't add to this conpleted PR.
